storage/backup/backup-cli/src/coordinators/restore.rs

```diff
@@ -88,110 +89,238 @@ impl RestoreCoordinator {
         ret
     }
 
-    async fn run_impl(mut self) -> Result<()> {
-        // N.b.
-        // The coordinator now focuses on doing one procedure, ignoring the combination of options
-        // supported before:
-        //   1. a most recent state snapshot before --target-version
-        //   2. a only transaction and its output, at the state snapshot version
-        //   3. the epoch history from 0 up until the latest closed epoch preceding the state
-        //      snapshot version.
-        // And it does so in a resume-able way.
-
+    /// Support two modes
+    /// 1. restore to target version
+    /// 2. restore a DB with all data ranging from start_version to target_version
+    /// We basically introduce a ledger history start version (lhs), a replay start version (rs) and target version
+    /// We directly store the write set and key values between (lhs, rs) and replay txn from (rs, target]
+    async fn run_impl(self) -> Result<()> {
         if self.replay_all {
             bail!("--replay--all not supported in this version.");
         }
 
+        info!("This tool only guarantees resume from previous in-progress restore. \
+        If you want to restore a new DB, please either specify a new target db dir or delete previous in-progress DB in the target db dir.");
+
         let metadata_view = metadata::cache::sync_and_load(
             &self.metadata_cache_opt,
             Arc::clone(&self.storage),
             self.global_opt.concurrent_downloads,
         )
         .await?;
 
-        let next_txn_version = self
+        let target_version = self.global_opt.target_version;
+        COORDINATOR_TARGET_VERSION.set(target_version as i64);
+
+        // calculate the start_version and replay_version
+        let max_txn_ver = metadata_view
+            .max_transaction_version()?
+            .ok_or_else(|| anyhow!("No transaction backup found."))?;
+        let lhs = self.ledger_history_start_version();
+
+        let latest_tree_version = self
+            .global_opt
+            .run_mode
+            .get_state_snapshot_before(Version::MAX);
+        let tree_completed = {
```

Comment:
nuh.. this doesn't mean the tree is complete.. just there is a tree. (see my previous comment -- you need to call the function again to see if there is only a (partial) tree at that version or there's another tree version before it -- in the latter case, we know trees are complete, while in the former case, the tree can be partial, and that's where you need to try to resume)